src/highdicom/sr/content.py

```diff
@@ -599,3 +838,163 @@ def __init__(
+        segment_number: int
+            number of the segment to reference within the provided dataset
+        frame_numbers: Optional[Sequence[int]], optional
+            list of frames in the segmentation dataset to reference. If
+            not provided, the reference is assumed to apply to all frames
+            of the given segment number
+
+        Returns
+        -------
+        ReferencedSegment
+            Content item representing a reference to the segment
+
+        Notes
+        -----
+             This method will attempt to deduce source image information
+             from information provided in the segmentation instance. If
+             available, it will used information specific to the segment
+             and frame numbers (if any) provided using the Derivation
+             Image Sequence information in the frames. If this information
+             is not present in the segmentation dataset, it will instead
+             use the information in the Referenced Series Sequence, which
+             applies to all segments and frames present in the segmentation
+             instance.
+
+        """
+        if dataset.SOPClassUID != SegmentationStorage:
+            raise ValueError(
+                'Input dataset should be a segmentation storage instance'
+            )
+
+        source_images = []
+        source_series = None
+
+        # Attempt to retrieve a list of source images for the specific segment
+        # and frames specified using information in the per-frame functional
+        # groups sequence
+        # This gives more specific information, but depends upon optional
+        # attributes so may fail with some segmentation objects
+        # Those created with highdicom should have this information
+        referenced_frame_info = []
+        if frame_numbers is not None:
+            # Use only the provided frames
+            for f in frame_numbers:
+                # Check that the provided frame number is valid
+                if f < 0 or f >= dataset.NumberOfFrames:
+                    raise ValueError(
+                        f'Frame {f} is an invalid frame number within the '
+                        'provided dataset'
+                    )
+
+                frame_info = dataset.PerFrameFunctionalGroupsSequence[f]
+
+                # Check that this frame references the correct
+                # segment
+                ref_segment = frame_info.SegmentIdentificationSequence[0]\
+                    .ReferencedSegmentNumber
+
+                if ref_segment != segment_number:
+                    raise ValueError(
+                        f'The provided frame number {f} does not refer to '
+                        f'segment number {segment_number}'
+                    )
+                referenced_frame_info.append(frame_info)
+        else:
+            # Use every frame that applies to this segment
+            for frame_info in dataset.PerFrameFunctionalGroupsSequence:
```

Comment:
Since there may be many frames in a Segmentation instance, I would rewrite this into a list comprehension for improved performance.

Reply:
Addressed in 4bea49d